lib/src/rules/avoid_unstable_final_fields.dart

```diff
@@ -0,0 +1,685 @@
+
+    bool containsNonConstantType(TypeAnnotation? typeAnnotation) {
+      if (typeAnnotation == null) return false;
+      if (typeAnnotation is NamedType) {
+        var typeArguments = typeAnnotation.typeArguments;
+        if (typeArguments != null) {
+          for (var typeArgument in typeArguments.arguments) {
+            if (containsNonConstantType(typeArgument)) return true;
+          }
+        }
+        var element = typeAnnotation.type?.element?.declaration;
+        if (element is ClassElement) return false;
+        if (element is TypeParameterElement) {
+          var owner = element.enclosingElement;
+          // A class type parameter is not constant, but it is stable.
+          if (owner is ClassElement) return false;
+          return true;
+        }
+        // TODO(eernst): Handle `typedef` and other missing cases.
+        return true;
+      } else if (typeAnnotation is GenericFunctionType) {
+        // TODO(eernst): For now, just use the safe approximation.
+        return true;
+      } else {
+        // TODO(eernst): Add missing cases. Be safe for now.
+        return true;
+      }
+    }
+
+    if (containsNonConstantType(node.type)) isStable = false;
+  }
+}
+
+class _FieldVisitor extends _AbstractVisitor {
+  _FieldVisitor(super.rule, super.context);
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    Uri? libraryUri;
+    Name? name;
+    ClassElement? classElement;
+    for (var variable in node.fields.variables) {
+      var declaredElement = variable.declaredElement;
+      if (declaredElement is FieldElement) {
+        // A final instance variable can never violate stability.
+        if (declaredElement.isFinal) continue;
+        // A non-final instance variable is always a violation of stability.
+        // Check if stability is required.
+        classElement ??= declaredElement.enclosingElement as ClassElement;
+        libraryUri ??= declaredElement.library.source.uri;
+        name ??= Name(libraryUri, declaredElement.name);
+        if (_inheritsStability(classElement, name, context)) {
+          doReportLint(node, variable.name);
+        }
+      }
+    }
+  }
+}
+
+class _MethodVisitor extends _AbstractVisitor {
+  late MethodDeclaration declaration;
+
+  _MethodVisitor(super.rule, super.context);
+
+  @override
+  void visitMethodDeclaration(MethodDeclaration node) {
+    if (!node.isGetter) return;
+    declaration = node;
+    var declaredElement = node.declaredElement;
+    if (declaredElement != null) {
+      var enclosingElement = declaredElement.enclosingElement;
+      if (enclosingElement is ClassElement) {
+        var libraryUri = declaredElement.library.source.uri;
+        var name = Name(libraryUri, declaredElement.name);
+        if (!_inheritsStability(enclosingElement, name, context)) return;
+        node.body.accept(this);
+        if (!isStable) doReportLint(node, node.name);
+      } else {
+        // Extensions cannot override anything.
```

Comment:
You'll need to add branches for `EnumElement` and `MixinElement` when you rebase.

Reply:
Having rebased, I get a lot of github action failures, in libraries that I haven't touched: https://github.com/dart-lang/linter/runs/7767039216?check_suite_focus=true.

Reply:
The github actions fail, reporting 'info - lib/src/rules/avoid_unstable_final_fields.dart:641:39 - 'name' is deprecated and shouldn't be used. Use name2 instead. Try replacing the use of the deprecated member with the replacement. - deprecated_member_use', in spite of the fact that `name2` does not exist.

Reply:
The class `MethodDeclaration` defines both `name` and `name2`:
```
  /// Return the name of the method.
  @Deprecated('Use name2 instead')
  SimpleIdentifier get name;

  /// Return the name of the method.
  Token get name2;
```
It's possible that it was in a bad state at one point, or that `node` has been promoted to some other type. If you update and it still isn't right, let me know. (Or if the "Got it" below means this issue has been resolved, ignore this comment.)

Reply:
OK, I now get no error that `name2` does not exist, but it is a `Token`, and `reportLint` wants an `AstNode` (which is the `SimpleIdentifier` that contains the name for other lint reports). How do I get hold of an `AstNode` for the name when I only have a `Token`?

Reply:
> How do I get hold of an `AstNode` for the name when I only have a `Token`?

You can't. You can use `reportLintForToken` for those cases.

> Are lint messages not supposed to have context information? Is there some other mechanism which is more appropriate?

Yes, lints can have context messages.

The `reportLint*` methods take an optional `contextMessages:` argument that will associate the context messages with the diagnostic.

Reply:
Ah, great, `reportLintForToken` was just what I needed!

For the context messages I'm already using that optional parameter, but I couldn't find any other way to _create_ a diagnostic message than via the constructor of `DiagnosticMessageImpl`, and that's where I need to import a `src` library. Is there another way?

Reply:
No, there isn't, though there ought to be. I guess the answer is that you have to have a src import for now.

Reply:
OK, thanks a lot! I've added `// ignore:implementation_imports` just before that import, hoping that this will suppress the failure in the github actions.

Reply:
Worked!